### PR TITLE
libkbfs: refactor block put logic into helper functions

### PR DESCRIPTION
**Pull request by strib**
This includes:
* Moving the parallel block put functions into block_util.go.
* Removing the BlockOps.Put() method (since the block journal can't use
  BlockOps, but still wants parallel puts).
* Change stallingBlockOps into stallingBlockServer (since some tests need to
  stall on Put, and now BlockOps.Put is gone).

In a future PR, I'll change the journal to use this.

Issue: KBFS-1275

### CODE REVIEW

File: libkbfs/block_util.go [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+
+func isRecoverableBlockError(err error) bool {
+	_, isArchiveError := err.(BServerErrorBlockArchived)
+	_, isDeleteError := err.(BServerErrorBlockDeleted)
+	_, isRefError := err.(BServerErrorBlockNonExistent)
+	_, isMaxExceededError := err.(BServerErrorMaxRefExceeded)
+	return isArchiveError || isDeleteError || isRefError || isMaxExceededError
+}
+
+// Put implements the BlockOps interface for BlockOpsStandard.
```
Comment (akalin-keybase) [inline libkbfs/block_util.go]: fix/remove comment

File: libkbfs/interfaces.go [hunk 2]
```diff
@@ -942,12 +942,6 @@ type BlockOps interface {
 	Ready(ctx context.Context, kmd KeyMetadata, block Block) (
 		id BlockID, plainSize int, readyBlockData ReadyBlockData, err error)
 
-	// Put stores the readied block data under the given block
-	// pointer (which belongs to the TLF with the given ID) on the
-	// server.
-	Put(ctx context.Context, tlfID TlfID, blockPtr BlockPointer,
```
Comment (akalin-keybase) [inline libkbfs/interfaces.go]: If we end up making quota reclamation talk to the real block server directly, we might need to add this back, or have some other way of distinguishing talking to the 'real' block server vs. the journal block server. But we can figure that out when we get there...

File: libkbfs/block_util_test.go [hunk 3]
```diff
@@ -0,0 +1,86 @@
+package libkbfs
+
+import (
+	"errors"
+	"testing"
+)
+
+func TestBlockUtilPutNewBlockSuccess(t *testing.T) {
+	mockCtrl, config, ctx := blockOpsInit(t)
```
Comment (akalin-keybase) [inline libkbfs/block_util_test.go]: since we only use the mock block server, may as well have an init/shutdown function that returns just that

File: libkbfs/folder_block_manager_test.go [hunk 4]
```diff
@@ -381,6 +381,8 @@ func TestQuotaReclamationDeletedBlocks(t *testing.T) {
 	}
 
 	// Stall the puts that comes as part of the sync call.
+	oldBServer := config2.BlockServer()
+	defer config2.SetBlockServer(oldBServer)
```
Comment (akalin-keybase) [inline libkbfs/folder_block_manager_test.go]: Why was this necessary?

File: libkbfs/test_stallers.go [hunk 5]
```diff
@@ -311,75 +309,74 @@ func runWithContextCheck(ctx context.Context, action func(ctx context.Context) e
...
+// corresponding staller is used to stall the operation.
+type stallingBlockServer struct {
 	stallOpName StallableBlockOp
 	// stallKey is a key for switching on/off stalling. If it's present in ctx,
 	// and equal to `true`, the operation is stalled. This allows us to use the
 	// ctx to control stallings
 	stallKey string
 	staller  staller
-	delegate BlockOps
+	delegate BlockServer
```
Comment (akalin-keybase) [inline libkbfs/test_stallers.go]: Since you're passing through most functions now, may be better to just embed `BlockServer` and remove the delegating functions. What do you think?

**Comment by akalin-keybase** [comment]
lgtm after questions and nits

File: libkbfs/interfaces.go (hunk 2 above)
Comment (strib) [inline libkbfs/interfaces.go]: Hmm, quota reclamation doesn't involve putting blocks, only removing references, which is a direct call to `MDServer`.  So I don't think having `MDOps.Put` would change things one way or the other.

File: libkbfs/folder_block_manager_test.go (hunk 4 above)
Comment (strib) [inline libkbfs/folder_block_manager_test.go]: Because otherwise `StateChecker` fails because it's not allowed to state check on something that's not a `BlockServerLocal`.

**Comment by strib** [comment]
Thanks!  Made the changes you suggested, will push when CI passes.

**Comment by strib** [comment]
CI failure is known (cc @jzila: [TestMultipleDeviceReaderOffThenOn](https://ci.keybase.io/job/kbfs-server/job/master/847/execution/node/158/log/)).  Merging.

File: libkbfs/interfaces.go (hunk 2 above)
Comment (akalin-keybase) [inline libkbfs/interfaces.go]: Ah, you're right! (I assume you meant BlockServer and BlockOps)

Comment (strib) [inline libkbfs/interfaces.go]: Derp, yeah.
